File: python/N-Queens II.py
class Solution(object):
    used = None
    answer = 0
    size = 0
    
    def dfs(self, row, col):
        for i in range(0, self.size):
            if self.used[row][i] and i != col:
                return False
            if self.used[i][col] and i != row:
                return False
        
        i = row - 1
        j = col - 1
        while i >= 0 and j >= 0:
            if self.used[i][j]:
                return False
            i -= 1
            j -= 1

        i = row + 1
        j = col + 1
        while i < self.size and j < self.size:
            if self.used[i][j]:
                return False
            i += 1
            j += 1

        i = row - 1
        j = col + 1
        while i >= 0 and j < self.size:
            if self.used[i][j]:
                return False
            i -= 1
            j += 1

        i = row + 1
        j = col - 1
        while i < self.size and j >= 0:
            if self.used[i][j]:
                return False
            i += 1
            j -= 1

        if row == (self.size - 1):
            self.answer += 1

        for i in range(0, self.size):
            self.used[row + 1][i] = True
            self.dfs(row + 1, i)
            self.used[row + 1][i] = False
            
    def totalNQueens(self, n):
        self.used = [[False for col in range(n + 1)] for row in range(n + 1)]
        self.size = n
        self.answer = 0
        for i in range(0, n):
            self.used[0][i] = True
            self.dfs(0, i)
            self.used[0][i] = False
        
        return self.answer

File: cpp/N-Queens II.cpp
class Solution {
public:
    vector<bool> column;
    vector<bool> rightTop;
    vector<bool> leftTop1;
    vector<bool> leftTop2;
    int size, answer;
    
    int totalNQueens(int n) {
        if (n == 1) {
            return 1;
        }
        size = n;
        answer = 0;
        for (int i = 0; i < n + 1; i++) {
            column.push_back(false);
            leftTop1.push_back(false);
            leftTop2.push_back(false);
        }
        for (int i = 0; i < (n-1) * 2 + 1; i++) {
            rightTop.push_back(false);
        }
        helper(-1, 0);
        return answer;
    }
    
    void helper(int x, int y) {
        if (x >= size - 1) {
            answer++;
            return;
        }
        
        for (int i = 0; i < size; i++) {
            if (column[i] || rightTop[x+1+i]) {
                continue;
            }
            if (i > x+1 && leftTop1[i-(x+1)]) {
                continue;
            }
            if (x+1 >= i && leftTop2[x+1-i]) {
                continue;
            }
        
            column[i] = true;
            rightTop[x+1+i] = true;
            if (i > x+1) {
                leftTop1[i-(x+1)] = true;
            } else {
                leftTop2[x+1-i] = true;
            }
            
            helper(x + 1, i);
            
            column[i] = false;
            rightTop[x+1+i] = false;
            if (i > x+1) {
                leftTop1[i-(x+1)] = false;
            } else {
                leftTop2[x+1-i] = false;
            }
        }
    }
};
